### PR TITLE
Download previously saved uberjar for E2E workflow

### PR DESCRIPTION
**Pull request by nemanjaglumac**
E2E tests require a full-fledged Metabase artifact in order to run.
We've been **always** building it from scratch so far, even when there were no source code changes in a PR! This is wasteful and slow.

This PR introduces a new logic that will find the previously saved uberjar by traversing the current commit parents and searching for the matching uberjar.

If it finds it, the build will be skipped and E2E tests will download and will use that uberjar directly.
If it cannot find the uberjar or it fails, `build` job will take over as a fallback mechanism and it's business as usual.

## Download Logic Explained
GitHub API allows to list all artifacts across all workflows in a repo.
https://docs.github.com/en/rest/actions/artifacts?apiVersion=2022-11-28#list-artifacts-for-a-repository

If you try this endpoint, you'll realize just how many artifacts we store
```json
{
	"total_count": 30242,
        "artifacts":[{...}, {...}, ...{}]
}
```

Lucky for us, it accepts a few query parameters that will help us in this quest.
- `name`: we know we're searching for either "metabase-oss-uberjar" or `metabase-ee-uberjar"
- `page`: Page number of the results to fetch.
- `per_page`: How many results do we want to fetch per page to increase our chance of finding the right artifact (max is 100)

The individual artifact has the following shape:
```json
{
  "id": 861710849,
  "node_id": "MDg6QXJ0aWZhY3Q4NjE3MTA4NDk=",
  "name": "metabase-ee-uberjar",
  "size_in_bytes": 335089591,
  "url": "https://api.github.com/repos/metabase/metabase/actions/artifacts/861710849",
  "archive_download_url": "https://api.github.com/repos/metabase/metabase/actions/artifacts/861710849/zip",
  "expired": false,
  "created_at": "2023-08-14T18:47:16Z",
  "updated_at": "2023-08-14T18:47:19Z",
  "expires_at": "2023-09-13T18:17:04Z",
  "workflow_run": {
    "id": 5858087729,
    "repository_id": 30203935,
    "head_repository_id": 30203935,
    "head_branch": "mblib-column-settings-round-trip",
    "head_sha": "4ba2debf861132965d30fa69b8493c86a36ba9d0"
  }
}
```

In order to get the `archive_download_url` we will match the `head_sha` to the parent commit SHA1. I.e. we filter the list of all possible artifacts by the artifact name and the parent's SHA1.

Once we know the `archive_download_url`, the only thing left to do is to download and store it using `curl`.

To save some time, we're mapping `download_uberjar` job outputs[^1] to:
1. `enterprise_download_link`
2. `oss_download_link`

which we later use in the`e2e-tests` job to download, store and unzip the uberjar directly where it's needed.

### tl;dr
- if changed files include [source files](https://github.com/metabase/metabase/blob/a0eb1e8bf69c0b7a8a1bce4f89c0b98dc1ade5a6/.github/file-paths.yaml#L80)
    - download should be skipped
    - build should run as it did before
- if changed files include [only E2E specs and helpers](https://github.com/metabase/metabase/blob/a0eb1e8bf69c0b7a8a1bce4f89c0b98dc1ade5a6/.github/file-paths.yaml#L74)
    - download should find previously built uberjar
    - build should be skipped

### Detailed
For this I created a separate PR that:
1. only changes an E2E spec
2. merges into this one (so that the diff shows only that one spec)

https://github.com/metabase/metabase/pull/33202

Let's examine what happened:
- It tried to find the matching uberjar that matches its parent commit SHA1
    - It searched recursively for 3 pages, listing 100 artifacts per page
- Since it couldn't find it, it moved to the next parent, and so on
- And then it repeated the process until it reached f9d328cb8f597e078594da9e374d01ed3cb2df4d, which was the commit previously merged into `master`
    - This commit triggered workflows that built Metabase uberjar

If we take a look at [the job logs](https://github.com/metabase/metabase/actions/runs/5902919492/job/16011822942?pr=33202#step:3:95), we'll see exactly what happened:
```
Didn't find the artifact for 'a0eb1e8bf69c0b7a8a1bce4f89c0b98dc1ade5a6' on the page: 1.
Didn't find the artifact for 'a0eb1e8bf69c0b7a8a1bce4f89c0b98dc1ade5a6' on the page: 2.
Didn't find the artifact for 'a0eb1e8bf69c0b7a8a1bce4f89c0b98dc1ade5a6' on the page: 3.
Switching to a new parent
Didn't find the artifact for 'c9a7b4617a330da87fd7aeefc50b3cb0bd6bcc30' on the page: 1.
Didn't find the artifact for 'c9a7b4617a330da87fd7aeefc50b3cb0bd6bcc30' on the page: 2.
Didn't find the artifact for 'c9a7b4617a330da87fd7aeefc50b3cb0bd6bcc30' on the page: 3.
Switching to a new parent
Didn't find the artifact for '54e18f6a[47](https://github.com/metabase/metabase/actions/runs/5902919492/job/16011822942?pr=33202#step:3:48)38472c53aff46210d286ce5fbccc28' on the page: 1.
Didn't find the artifact for '54e18f6a4738472c53aff46210d286ce5fbccc28' on the page: 2.
Didn't find the artifact for '54e18f6a4738472c53aff46210d286ce5fbccc28' on the page: 3.
Switching to a new parent
Didn't find the artifact for '0fde8ca31cf39823592a95dad35bdbac7e6e6126' on the page: 1.
Didn't find the artifact for '0fde8ca31cf39823592a95dad35bdbac7e6e6126' on the page: 2.
Didn't find the artifact for '0fde8ca31cf39823592a95dad35bdbac7e6e6126' on the page: 3.
Switching to a new parent
Didn't find the artifact for '[48](https://github.com/metabase/metabase/actions/runs/5902919492/job/16011822942?pr=33202#step:3:49)97a17e385ca23e[49](https://github.com/metabase/metabase/actions/runs/5902919492/job/16011822942?pr=33202#step:3:50)f3d1691dba7d47081d2ece' on the page: 1.
Didn't find the artifact for '4897a17e385ca23e49f3d1691dba7d47081d2ece' on the page: 2.
Didn't find the artifact for '4897a17e385ca23e49f3d1691dba7d47081d2ece' on the page: 3.
Switching to a new parent
Didn't find the artifact for '4a14108d3a76a1e176e7b997ba7cee28a00e[51](https://github.com/metabase/metabase/actions/runs/5902919492/job/16011822942?pr=33202#step:3:52)a9' on the page: 1.
Didn't find the artifact for '4a14108d3a76a1e176e7b997ba7cee28a00e51a9' on the page: 2.
Didn't find the artifact for '4a14108d3a76a1e176e7b997ba7cee28a00e51a9' on the page: 3.
Switching to a new parent
Didn't find the artifact for 'ca1eeb167242e339a23c0cd0202b9e0ca262ab23' on the page: 1.
Didn't find the artifact for 'ca1eeb167242e339a23c0cd0202b9e0ca262ab23' on the page: 2.
Didn't find the artifact for 'ca1eeb167242e339a23c0cd0202b9e0ca262ab23' on the page: 3.
Switching to a new parent
Didn't find the artifact for '205bd67705ad1f06f798057d34c456cea355ca[52](https://github.com/metabase/metabase/actions/runs/5902919492/job/16011822942?pr=33202#step:3:53)' on the page: 1.
Didn't find the artifact for '205bd67705ad1f06f798057d34c456cea355ca52' on the page: 2.
Didn't find the artifact for '205bd67705ad1f06f798057d34c456cea355ca52' on the page: 3.
Switching to a new parent
Didn't find the artifact for '18eb09bed831faabbd45e6d2[53](https://github.com/metabase/metabase/actions/runs/5902919492/job/16011822942?pr=33202#step:3:54)88c7ded463c998' on the page: 1.
Didn't find the artifact for '18eb09bed831faabbd45e6d25388c7ded463c998' on the page: 2.
Didn't find the artifact for '18eb09bed831faabbd45e6d25388c7ded463c998' on the page: 3.
Switching to a new parent
Didn't find the artifact for '9da4fe4bda31641080bb0[54](https://github.com/metabase/metabase/actions/runs/5902919492/job/16011822942?pr=33202#step:3:55)2ae82ad743e8f6c1a' on the page: 1.
Didn't find the artifact for '9da4fe4bda31641080bb0542ae82ad743e8f6c1a' on the page: 2.
Didn't find the artifact for '9da4fe4bda31641080bb0542ae82ad743e8f6c1a' on the page: 3.
Switching to a new parent
Didn't find the artifact for 'ff1590dc7d61a3caad7f029[57](https://github.com/metabase/metabase/actions/runs/5902919492/job/16011822942?pr=33202#step:3:58)98c4f8a901f605d' on the page: 1.
Didn't find the artifact for 'ff1590dc7d61a3caad7f0295798c4f8a901f605d' on the page: 2.
Didn't find the artifact for 'ff1590dc7d61a3caad7f0295798c4f8a901f605d' on the page: 3.
Switching to a new parent
Didn't find the artifact for '6ced4c9e662da3a126bbda[59](https://github.com/metabase/metabase/actions/runs/5902919492/job/16011822942?pr=33202#step:3:60)14ed78089be51f0c' on the page: 1.
Didn't find the artifact for '6ced4c9e662da3a126bbda5914ed78089be51f0c' on the page: 2.
Didn't find the artifact for '6ced4c9e662da3a126bbda5914ed78089be51f0c' on the page: 3.
Switching to a new parent
Found metabase-ee uberjar for 'f9d328cb8f597e078594da9e3[74](https://github.com/metabase/metabase/actions/runs/5902919492/job/16011822942?pr=33202#step:3:75)d01ed3cb2df4d'!
```

[^1]: https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs

### CODE REVIEW

File: .github/file-paths.yaml [hunk 1]
```diff
@@ -80,7 +80,6 @@ e2e_specs: &e2e_specs
 e2e_all:
   - *default
   - *ci
-  - *e2e_specs
```
Comment (nemanjaglumac) [inline .github/file-paths.yaml:83]: The gist of this change is to decouple pure E2E spec and helper changes from all other (source code) changes that will trigger the actual build process.

File: .github/workflows/e2e-tests.yml [hunk 2]
```diff
@@ -38,9 +39,79 @@ jobs:
...
+          current_commit="HEAD"
+
+          get_parent () {
+            parent_commit=$(git rev-parse $1^)
+            echo $parent_commit
+          }
+
+          get_download_link () {
+            parent_commit=$(get_parent $1)
+            current_page=${2:-1}
```
Comment (nemanjaglumac) [inline .github/workflows/e2e-tests.yml:72]: Optional second argument with the default value
https://stackoverflow.com/a/9333006/8815185

**Comment by qnkhuat** [comment]
Have you thought about rate limit from GH APIs?

I wonder how many APIs we are calling per commit and what's our rate limit?

Ref: https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#rate-limits-for-requests-from-github-actions

**Review by iethree — APPROVED** [review]
🏅 Looks good to me! I think this will be useful on a relatively small percentage of PRs, but will be really helpful when developing tests!

**Comment by nemanjaglumac** [comment]
> Have you thought about rate limit from GH APIs?

Yes I have.
We're doing this authenticated, and I've adjusted the number of retries to account for the rate API limit. We should be good.
